File: apps/api/src/ichor_api/services/recent_actuals.py
"""recent_actuals -- query economic_events with published `actual` + classify.

r145 (ADR-099 §Impl) -- closes Mission centrale axis-5 user-surface visibility :
r144 lit the `actual` column for US tier-1 events via FRED ALFRED ; r145
surfaces those rows on `/briefing/[asset]` with the r141 5-state classifier
wired as the API truth-source.

Pure compute + ORM read. Wires `classify_surprise()` (dormant since r141)
as the single API truth-source : today `state=unavailable` for all events
(no analyst range provider live yet) but `magnitude_pct` is computed via
the consensus point estimate (FF `forecast` column), preserving useful
"how far from consensus" info. When the r146+ range provider lands, state
badges auto-light up without API/frontend changes.

ADR-017 compliance : the classifier produces descriptive geometric labels
(`above_range`, `below_range`, `in_range`, `exact_consensus`, `unavailable`)
+ signed `magnitude_pct` -- never directional. Per-asset transmission lives
in verdict/confluence layers (see `MacroSurprisePanel.tsx:151-154` doctrine).

Doctrine #11 calibrated honesty : `unavailable` rows surface raw values
silently (a11y aria-label only -- no fabricated badge). Lesson #37 :
DEMOTE framing when upstream data lacks actionable field (range provider
not yet wired -> classifier state cannot fire -> magnitude only).

ADR refs : ADR-099 §Impl(r145) -- Mission centrale Axis-5 visible surface.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..models import EconomicEvent
from .economic_event_surprise import SurpriseClassification, classify_surprise

__all__ = [
    "RecentActualRow",
    "fetch_recent_actuals",
]


@dataclass(frozen=True)
class RecentActualRow:
    """A single past economic event with published `actual` + classification.

    Pairs the raw text columns (display-ready, no fabrication when None)
    with the `classify_surprise()` result. Both layers are exposed so the
    frontend can show raw values AND the classifier verdict side-by-side
    (parity with `MacroSurprisePanel` raw z + magnitude pattern).

    Attributes :
        event_id        : UUID stringified -- React key, never displayed.
        currency        : 3-letter ISO ("USD", "EUR", etc.).
        scheduled_at    : event firing time (UTC, ISO 8601) ; never None
                          here because we filter `scheduled_at IS NOT NULL`.
        title           : ForexFactory event title verbatim.
        impact          : FF impact tier ("high", "medium", "low").
        actual          : raw text from DB, never None here (filtered).
        forecast        : raw consensus point text (may be None).
        forecast_min    : raw envelope lower bound (NULL today, r146+).
        forecast_max    : raw envelope upper bound (NULL today, r146+).
        previous        : raw prior period value text (may be None).
        url             : ForexFactory event detail URL (may be None).
        classification  : `SurpriseClassification` from r141 classifier.
    """

    event_id: str
    currency: str
    scheduled_at: datetime
    title: str
    impact: str
    actual: str
    forecast: str | None
    forecast_min: str | None
    forecast_max: str | None
    previous: str | None
    url: str | None
    classification: SurpriseClassification


async def fetch_recent_actuals(
    session: AsyncSession,
    *,
    lookback_days: int = 30,
    currency: str | None = "USD",
    limit: int = 25,
    now: datetime | None = None,
) -> list[RecentActualRow]:
    """Fetch past N-day economic events with published `actual` + classify.

    Args :
        session        : SQLAlchemy async session.
        lookback_days  : how many days back to scan (default 30 -- matches
                         r144 reconciler default window).
        currency       : ISO currency filter ("USD" default to match the
                         only populated bucket today via r144). Pass None
                         to skip filter (returns all currencies).
        limit          : cap rows returned (default 25, max 200). Newest
                         events first by `scheduled_at DESC`.
        now            : injection point for deterministic testing. Default
                         `datetime.now(UTC)`.

    Returns rows newest-first. Each row carries both raw text + classifier
    verdict so the frontend can render raw values + magnitude_pct token
    even when `state=unavailable` (no range data yet).

    Never raises -- empty list on no matches. ORM query only, no I/O.
    """
    if now is None:
        now = datetime.now(UTC)
    if limit < 0:
        limit = 0
    # Hard cap defensive against caller passing huge `limit` via query
    # string -- /v1 router will pre-validate via `Query(le=200)` but the
    # service layer also clamps as belt-and-suspenders (parity with
    # `economic_event_actuals_reconciler` r144 defensive clamps).
    limit = min(limit, 200)
    lookback_floor = now - timedelta(days=max(lookback_days, 0))

    stmt = (
        select(EconomicEvent)
        .where(
            EconomicEvent.scheduled_at.is_not(None),
            EconomicEvent.scheduled_at >= lookback_floor,
            EconomicEvent.scheduled_at <= now,
            EconomicEvent.actual.is_not(None),
        )
        .order_by(EconomicEvent.scheduled_at.desc())
        .limit(limit)
    )
    if currency is not None:
        stmt = stmt.where(EconomicEvent.currency == currency)

    rows = (await session.execute(stmt)).scalars().all()

    result: list[RecentActualRow] = []
    for r in rows:
        # Defensive null-guard -- filtered above but mypy can't see the SQL
        # predicate. If schema invariants ever weaken, the runtime check
        # keeps us honest rather than emitting a half-broken row.
        if r.scheduled_at is None or r.actual is None:
            continue
        classification = classify_surprise(
            actual=r.actual,
            consensus=r.forecast,
            forecast_min=r.forecast_min,
            forecast_max=r.forecast_max,
        )
        result.append(
            RecentActualRow(
                event_id=str(r.id),
                currency=r.currency,
                scheduled_at=r.scheduled_at,
                title=r.title,
                impact=r.impact,
                actual=r.actual,
                forecast=r.forecast,
                forecast_min=r.forecast_min,
                forecast_max=r.forecast_max,
                previous=r.previous,
                url=r.url,
                classification=classification,
            )
        )
    return result

File: apps/web2/lib/recentActuals.ts
/**
 * lib/recentActuals.ts — pure-fn view-model for the r145 RecentActualsPanel
 * (Mission centrale axis-5 visible surface).
 *
 * Surfaces past economic events that have a published `actual` (r144 FRED
 * ALFRED reconciler populates this column for US tier-1 events) classified
 * via r141 `classify_surprise`. The backend wires the classifier as the
 * single API truth-source ; this module is the pure-fn FR copy + magnitude
 * band layer.
 *
 * WHAT THE 5-STATE MEANS (researcher R59 r145 + verbatim Eliot transcript
 * codified r141 :
 *   *"si on sort à 3 % alors oui... ça restait dans le range, ça va pas
 *    surprendre. Alors que si on sort à 3.2 là ça vient vraiment changer
 *    la donne."* )
 *
 * The vocabulary is GEOMETRIC (actual vs envelope position), NOT directional
 * (bullish/bearish on the asset). A hot CPI is `above_range` here — whether
 * that's USD-bullish depends on the regime layer (counter-intuitive in
 * late-cycle "bad news is good news" regime — researcher §5).
 *
 * r145 reality : the analyst range envelope provider is not yet wired
 * (r146+ candidate). All rows return `state=unavailable` today. But
 * `magnitude_pct` IS populated from the FF consensus point, so we render
 * the magnitude token + raw values even when no state badge fires.
 *
 * ADR-017 boundary : DESCRIPTIVE geometric labels + signed scalars only.
 * The `unavailable` state surfaces row content silently (a11y aria-label
 * only, no fabricated badge — doctrine #11 calibrated honesty parity with
 * `economic_event_surprise.py:29-32`).
 *
 * Pure-fn module — RSC-safe, no React, no I/O.
 */

import type { RecentActualRow, SurpriseState } from "@/lib/api";

/** FR copy per state, locked by R59-researcher (r145).
 *
 * - `unavailable` : honest absence (range provider not live yet).
 * - `in_range` : Eliot transcript verbatim "*restait dans le range*".
 * - `above_range` / `below_range` : geometric distance, NOT a value
 *   judgement.
 * - `exact_consensus` : conversational + precise. Note "au point médian"
 *   would be technically wrong (consensus point ≠ midpoint of range).
 *
 * CI-pinned by `test_invariants_ichor.py` (r145 extension) — these strings
 * MUST stay `is_adr017_clean()`. */
export const SURPRISE_STATE_FR: Record<SurpriseState, string> = {
  unavailable: "Donnée non publiée",
  in_range: "Dans la fourchette des analystes",
  above_range: "Au-dessus de la fourchette",
  below_range: "En-dessous de la fourchette",
  exact_consensus: "Pile sur le consensus",
};

/** Threshold above which |magnitude_pct| is rendered in amber (notable
 * deviation from consensus). Mirrors the MacroSurprisePanel |z| ≥ 2
 * discipline — large surprise = worth attention, NOT good/bad. The cut
 * is intentionally conservative (5% pct deviation = roughly the line
 * between "noise" and "the market notices" for FF tier-1 macro). */
export const NOTABLE_MAGNITUDE_PCT_THRESHOLD = 5.0;

/** Format magnitude_pct as a compact signed text token : "+5.2%" / "−1.8%"
 * / "n/a". Polarity-neutral geometric framing — no "beat" / "miss" /
 * directional vocabulary.
 *
 * ui-designer r145 I1 : prior version included " vs consensus" inline (19
 * chars, broke 320px row layout). Suffix moved to the footer caveat band
 * + aria-label-only context. Token is now ~6 chars, parity with
 * MacroSurprisePanel's `+1.8σ` width. */
export function fmtMagnitudePct(pct: number | null): string {
  if (pct === null || !Number.isFinite(pct)) return "n/a";
  const sign = pct >= 0 ? "+" : "−";
  const abs = Math.abs(pct);
  return `${sign}${abs.toFixed(1)}%`;
}

/** Tone CSS variable for a magnitude_pct value, gated on classifier state.
 *
 * CONCORDANT 2/4 fix (ui-designer I2 + a11y SHOULD-1) : reserve amber
 * (`--color-warn`) ONLY for rows where the classifier state is meaningful
 * (i.e. state ≠ "unavailable"). r145 reality : no range provider yet means
 * EVERY row is `unavailable` today, so no amber appears -- the magnitude
 * we show is computed from the point forecast only, NOT a verified
 * "above-range breach". Surfacing amber on a half-known signal would be
 * fabricated emphasis.
 *
 * When the r146+ range provider lands, state badges auto-light up AND the
 * amber emphasis fires concordantly. */
export function magnitudePctTone(pct: number | null, stateMeaningful: boolean = false): string {
  if (pct === null || !Number.isFinite(pct)) return "var(--color-text-muted)";
  if (stateMeaningful && Math.abs(pct) >= NOTABLE_MAGNITUDE_PCT_THRESHOLD) {
    return "var(--color-warn)";
  }
  return "var(--color-text-secondary)";
}

/** Whether to render a state badge for this row.
 *
 * Researcher R59 §5 : `unavailable` MUST render the row silently with
 * `aria-label` only — no fabricated fallback. Today (r145) ALL rows are
 * `unavailable` (no range provider) so no badges fire ; tomorrow (r146+)
 * when the range lands, badges auto-light up.
 *
 * Note this returns true for `exact_consensus` (it's a meaningful state
 * even though it's geometrically "between"). */
export function shouldRenderStateBadge(state: SurpriseState): boolean {
  return state !== "unavailable";
}

/** Format the scheduled_at ISO timestamp for tile display.
 *
 * Returns a short HH:MM Paris-local string (we always render briefing in
 * Paris-tz per project convention). Date is implied by section header
 * "30 derniers jours" so we don't repeat it per row.
 *
 * Defensive : returns "—" on malformed input (RecentActualRow.scheduled_at_utc
 * is ISO 8601 per backend contract, but the wire shape may surprise). */
export function fmtScheduledAtParis(iso: string): string {
  try {
    const d = new Date(iso);
    if (Number.isNaN(d.getTime())) return "—";
    return d.toLocaleTimeString("fr-FR", {
      hour: "2-digit",
      minute: "2-digit",
      timeZone: "Europe/Paris",
    });
  } catch {
    return "—";
  }
}

/** Format the date part (DD/MM) for tile display — used as a secondary
 * line under the time to disambiguate when multiple events span days. */
export function fmtScheduledDateParis(iso: string): string {
  try {
    const d = new Date(iso);
    if (Number.isNaN(d.getTime())) return "";
    return d.toLocaleDateString("fr-FR", {
      day: "2-digit",
      month: "2-digit",
      timeZone: "Europe/Paris",
    });
  } catch {
    return "";
  }
}

/** Pure-fn empty-check : returns true when no rows at all. Used by the
 * panel to render honest silent absence (doctrine #11 — never fabricate
 * empty-state copy that pretends data exists). */
export function isEmptyRecentActuals(
  rows: ReadonlyArray<RecentActualRow> | null | undefined,
): boolean {
  return !rows || rows.length === 0;
}
